### PR TITLE
[SPARK-29160][CORE] Use UTF-8 explicitly for reading/writing event log file

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

Credit to @vanzin as he found and commented on this while reviewing #25670 - [comment](https://github.com/apache/spark/pull/25670#discussion_r325383512).

This patch proposes to specify UTF-8 explicitly while reading/writer event log file.

### Why are the changes needed?

The event log file is being read/written as default character set of JVM process which may open the chance to bring some problems on reading event log files from another machines. Spark's de facto standard character set is UTF-8, so it should be explicitly set to.

### Does this PR introduce any user-facing change?

Yes, if end users have been running Spark process with different default charset than "UTF-8", especially their driver JVM processes. No otherwise.

### How was this patch tested?

Existing UTs, as ReplayListenerSuite contains "end-to-end" event logging/reading tests (both uncompressed/compressed).

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
As I commented in `Does this PR introduce any user-facing change?` section, it may not be backward compatible change for some users. Please take this into consideration. Thanks!

**Comment by HeartSaVioR** [comment]
We might be able to remedy the backward incompatible change via having new option to let ReplayListenerBus use default character set to read file, though I'm not 100% sure it's a good workaround.

**Review by HyukjinKwon — APPROVED** [review]
Looks good but let me leave it to @vanzin

**Comment by HyukjinKwon** [comment]
@HeartSaVioR, if you concern about compatibility, you could leave a note in migration guide at https://github.com/apache/spark/blob/master/docs/core-migration-guide.md . I guess most of machines use UTF-8 by default though.

**Comment by dongjoon-hyun** [comment]
+1 for the migration note. cc @gatorsmile .
Also, cc @MaxGekk

**Comment by gatorsmile** [comment]
+1 to document it in the migration note.

**Comment by viirya** [comment]
Should we provide a config for this? I am not sure if we state in migration guide this change, how end users can react for it, if they want to use previous character set? Or this is very rare and we can ignore?

**Comment by HeartSaVioR** [comment]
There're three ways to deal with this:

1) provide simple tool to read input file with given character set and rewrite to UTF-8
2) expose new config to allow custom character set (or even just use default character set) on ReplayListenerBus
3) don't deal with this

I guess Spark hasn't provide additional tool to migrate existing one so 1) doesn't sound preferred one. If we would concern about backward compatibility, 2) seems to be the only option - remaining one is to allow custom character set or to use default charset.

**Comment by HeartSaVioR** [comment]
Btw, I could find another spots creating PrintWriter without explicit character set. Haven't touched them as I'd like to restrict/minimize the effects of changes.

**Comment by HeartSaVioR** [comment]
Build failure seems to be flaky one: `Build step 'Execute shell' marked build as failure`

**Comment by MaxGekk** [comment]
`UFT-8` -> `UTF-8` in the description: _Spark's de facto standard character set is **UFT-8**, so it should be explicitly set to_

File: core/src/main/scala/org/apache/spark/scheduler/ReplayListenerBus.scala [hunk 1]
```diff
@@ -54,7 +55,7 @@ private[spark] class ReplayListenerBus extends SparkListenerBus with Logging {
       sourceName: String,
       maybeTruncated: Boolean = false,
       eventsFilter: ReplayEventsFilter = SELECT_ALL_FILTER): Unit = {
-    val lines = Source.fromInputStream(logData).getLines()
+    val lines = Source.fromInputStream(logData, StandardCharsets.UTF_8.name()).getLines()
```
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/scheduler/ReplayListenerBus.scala]: You could avoid converting the charset to string and looking it up again by:
```scala
val lines = Source.fromInputStream(logData)(Codec.UTF8).getLines()
```

**Comment by HyukjinKwon** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/ReplayListenerBus.scala (hunk 1 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/scheduler/ReplayListenerBus.scala]: Nice suggestion! Will address.

**Comment by HeartSaVioR** [comment]
Build failure: known flaky test - https://issues.apache.org/jira/browse/SPARK-25903

**Comment by HeartSaVioR** [comment]
retest this, please

**Review by srowen — COMMENTED** [review]
Looks good in principle

**Review by MaxGekk — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
Uh, could we talk about which option is preferred one? It needs to be done before describing in migration note.

**Comment by HyukjinKwon** [comment]
I think we can just leave it without other options. It's rather corner case and I think it's fine to break such stuff since we're moving to Spark 3.

**Comment by viirya** [comment]
Ok for me.

**Review by viirya — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
Great :) Please let me know if someone would still want to provide backward compatible option. (Even it can be addressed separately.)

**Comment by dongjoon-hyun** [comment]
For me, the only remaining thing is a note at the migration guide to give prior warning to users, isn't it? It would be great if this PR includes that.
- https://github.com/apache/spark/pull/25845#issuecomment-532971687
- https://github.com/apache/spark/pull/25845#issuecomment-532987837
- https://github.com/apache/spark/pull/25845#issuecomment-532994719

**Comment by HeartSaVioR** [comment]
Yeah I thought someone may say it should worth to provide backward-compatible option so waited for couple of days to see more voices, but doesn't look like so. I'll mention the change to the migration note.

**Comment by dongjoon-hyun** [comment]
Ya. I agree with you.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by kiszk — APPROVED** [review]
nit: 'Spark History Server' -> 'Spark History Server (SHS)'

Is SHS commonly used in the community?

**Comment by HeartSaVioR** [comment]
Just rephrased SHS to Spark History Server as I guess it's official term and more commonly used. I guess SHS is also widely used term, but it might be a view of one of contributors, not one of end users.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Pull request closed without merging**